src/rulesets/LinearAlgebra/symmetric.jl

```diff
@@ -279,6 +279,172 @@ function rrule(::typeof(svdvals), A::LinearAlgebra.RealHermSymComplexHerm{<:BLAS
     return S, svdvals_pullback
 end
 
+#####
+##### matrix functions
+#####
+
+# Formula comes from so-called Daleckiĭ-Kreĭn theorem originally due to
+# Ju. L. Daleckiĭ and S. G. Kreĭn. Integration and differentiation of functions of Hermitian
+# operators and applications to the theory of perturbations.
+# Amer. Math. Soc. Transl., Series 2, 47:1–30, 1965.
+# Stabilization for almost-degenerate matrices due to
+# S. D. Axen, 2020. Representing Ensembles of Molecules.
+# Appendix D: Automatic differentation rules for power series functions of diagonalizable matrices
+# https://escholarship.org/uc/item/6s62d8pw
+# These rules are more stable for degenerate matrices than applying the chain rule to the
+# rules for `eigen`.
+
+for func in (:exp, :log, :sqrt, :cos, :sin, :tan, :cosh, :sinh, :tanh, :acos, :asin, :atan, :acosh, :asinh, :atanh)
+    @eval begin
+        function frule((_, ΔA), ::typeof($func), A::LinearAlgebra.RealHermSymComplexHerm)
+            ΔA isa AbstractZero && return $func(A), ΔA
+            Y, cache = _matfun($func, A)
+            Ȳ = _matfun_frechet($func, A, Y, ΔA, cache)
+            # If ΔA was hermitian, then ∂Y has the same structure as Y
+            ∂Y = if ishermitian(ΔA) && (isa(Y, Symmetric) || isa(Y, Hermitian))
+                _symhermlike!(Ȳ, Y)
+            else
+                Ȳ
+            end
+            return Y, ∂Y
+        end
+
+        function rrule(::typeof($func), A::LinearAlgebra.RealHermSymComplexHerm)
+            Y, cache = _matfun($func, A)
+            $(Symbol("$(func)_pullback"))(ΔY::AbstractZero) = (NO_FIELDS, ΔY)
+            function $(Symbol("$(func)_pullback"))(ΔY)
+                # for Hermitian Y, we don't need to realify the diagonal of ΔY, since the
+                # effect is the same as applying _hermitrize! at the end
+                ∂Y = eltype(Y) <: Real ? real(ΔY) : ΔY
+                # for matrix functions, the pullback is related to the pushforward by an adjoint
```

Comment:
No, a different reason. I don't have a name for the property, but is a property of any function that can be written as a converging power series with real coefficients. I haven't posted the proof anywhere or seen it before, but I'm sure it could be argued from some generic property. For Hermitian matrices, it can be shown by applying the usual inner product trick in the ChainRules docs to derive the pullback.

Reply:
Perhaps @antoine-levitt knows of this property.

Reply:
sorry, what property are you talking about?

Reply:
> sorry, what property are you talking about?

For a matrix function `f` that can be written in terms of a converging power series with real coefficients,
its pullback `f^*` is related to its pushforward `f_*` by an adjoint. Specifically, if `Y=f(A)`, `ΔA` is a tangent of `A`, and `ΔY` is a cotangent of `Y` (adopting ChainRules'/Zygote's conventions for how a cotangent is represented), then `(f^*)_{Y} (ΔY) = (f_*)_{A'} (ΔY)`. The property means we can write the pullback for any matrix function in terms of its `frule`.

Reply:
I'd have to check but isn't it basically f(A') = f(A)'?

Reply:
I got there through a much more complicated route (writing the power series as a recurrence and working out the corresponding pushforwards and pullbacks as recurrences yields this relation), but I'll check to see if we can get there from that property as well. That would be nice, haha.

Reply:
hm no it looks more complicated than that, it's df(A)' = df(A') or something like that, I'm not sure it follows from f(A)' = f(A') (but maybe it does, haven't checked carefully). I use this trick in the hermitian case because it means that the differential is self-adjoint on the space of hermitian matrices equipped with the Frobenius metric, but I didn't know about it in the non-hermitian case, it's cute.

Reply:
Seems like it would be worth me writing this up somewhere.

Reply:
OK, so let's do the proof for A^n, n ≥ 1. Then we can pass to arbitrary functions by linearity.

df(A)⋅dA = ∑_{k=0^n-1} A^k dA A^(n-1-k), so let's look at the adjoint of the linear operator L(A) : dA -> A^k1 dA A^k2, and the result follows again by linearity
<L(A) dA, dB> = tr((L(A) dA)' dB) = tr(A^k2' dA' A^k1' dB) = tr(dA' A^k1' dB A^k2') so adj(L(A)) : dB -> A^k1' dB A^k2' so adj(L(A)) = L(A')

That's probably close to the derivation you had?

Reply:
Okay yes, but yours is _much_ simpler and more concise than mine. Nice!